doc/memos/rdm-draft-petry-riot-design-goals.md

```diff
@@ -0,0 +1,345 @@
+modules, but they shouldn't be unnecessarily fine-grained either.
+
+## Cross-hardware portability
+
+Users might want to write a program for one piece of hardware, and later run it
+on another. RIOT should let user code be completely portable, so long as it's
+valid with the hardware.
+
+The hardware abstraction layer should be stable, well defined and consistent.
+Above the HAL, the only thing that modules should know about hardware is
+whether a required feature is there or not. If it isn't, the module should
+adapt accordingly, or not compile.
+
+## Real-time capabilities
+
+Different real-time guarantees are required for different use cases. Low
+frequency sensing needs only soft real-time support and can handle less timing
+accuracy so long as the timers support long timescales. High-frequency sensing
+and control applications need hard real-time guarantees and sub-millisecond
+timing accuracy.
+
+RIOT should deliver soft real-time performance by default, to cover the widest
+range of use cases. Programs with hard deadlines should be achievable with more
+expertise. It should provide timers which can competently handle the timescales
+of any application.
+
+## Interoperability
+
+RIOT nodes need to communicate reliably with non-RIOT nodes. The protocols need
+to be familiar to them and behave in the way they expect.
+
+RIOT should support open network protocol standards once they have reached a
+certain level of validity and popularity. It should be configurable so that
+users can choose which (optional) areas of specifications they want to include.
+Whatever the configuration, nodes should handle all possible traffic in a
+compliant manner.
+
+## Stability
+
+Nodes shouldn't fail because of RIOT, whatever the platform. Assuming perfect
+hardware and a moderately well-written application, they should be able to run
+indefinitely.
+
+The testing and peer review processes that ensure this should be under
+continuous review and refinement, including automating them where it's
+possible. Error handling should guarantee stability with minimal memory usage.
+
+## Unified APIs
+
+Users don't want to be confused by semantic differences between APIs. RIOT
+provides a similar "look and feel" across all our APIs.
+
+RIOT's interfaces should only differ in the ways that their modules differ.
+Where possible, layers should exist that sit on top of a certain class of
+modules to give an identical interface to the user. Semantic and naming
+conventions should be consistent throughout the system.
+
+## Internet security
+
+RIOT nodes need to be resilient against cyber attacks, for the sake of its
+reputation and the reputation of the IoT.
+
+It should be hard against all the threats a node is likely to experience,
+depending on the situation, including those from computers with many times
+their processing power. Security should be as easy to use as possible while
+still strong. Security flaws should be patched as quickly as possible.
```

Comment:
Please add:
"Security flaws should be patched as quickly as possible" + "by servicing a security alert channel with high priority."